### ISSUE TITLE
Feature request: Disallow returning a void-returning function result

### ISSUE THREAD

**Issue by d-luk** [typescript-eslint/typescript-eslint#2425]
I've seen the following pattern in code several times now, and it leads to some confusion:

```ts
function doSomething(): void {
    // Returns nothing
}

const result = await myPromise.then(() => {
    return doSomething();
});
```

The bottom statement suggests that a value is being returned from `doSomething()`, while in reality its type is `void`.

It would be great to have a rule that enforces removing this misleading `return` keyword:

```ts
const result = await myPromise.then(() => {
    doSomething();
});
```

This code style removes the suggestion that `result` will be anything other than `void` / `undefined`.

I think this rule is a good fit for `@typescript-eslint/eslint-plugin` because it must know about the return type of the called function.

**Comment by papb** [comment]
Should this rule disallow assigning `void` to a variable as well? For example, in your code, why would you want this `const result =`?

**Comment by d-luk** [comment]
@papb Good point. I don't see a reason for assigning `void` to a variable, so I'd say this should be disallowed as well.

**Comment by papb** [comment]
Nice. How about `no-void-assignment`? Disallowing `void` from assignments and return statements. I'd probably use enable that rule.

**Comment by d-luk** [comment]
Sounds good to me. 

Although other people might like to have the option to configure `no-void-assignment` and `no-void-return` separately, so perhaps there should be an option for that. I'd use both though.

**Comment by phaux** [comment]
How about [`no-void-expression`](https://palantir.github.io/tslint/rules/no-void-expression/) from TSLint, which forbids `void`-returning expressions everywhere except in statement position.

I'd like to try porting this rule.
